platform/logger_internal.h

```diff
@@ -0,0 +1,284 @@
+
+#define MBED_LOG_VSTR_2(fmt1, ap)                     log_str_vdata(fmt1, ap)
+#define MBED_LOG_VSTR_1(ll, mod, fmt, ap)             MBED_LOG_VSTR_2("[" ll "][" mod "]: " fmt, ap)
+
+/*  Internal function for assertion
+ *
+ *  This should be used to log assert information, application will termiate post logging
+ *
+ *  @param format    printf like format
+ *  @param ...       variable arguments related to format
+ *  @note: Application will terminate after printing assert log.
+ */
+#if defined(__GNUC__) || defined(__CC_ARM)
+void log_assert(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+#else
+void log_assert(const char *format, ...);
+#endif
+
+/*  Internal binary data logging function.
+ *
+ *  This should be used to log binary data.
+ *
+ *  @param argCount  Count of following variable arguments
+ *  @param args      variable arguments (uint32_t)
+ *  @note Only uint32_t type variable arguments accepted.
+ */
+void log_id_data(uint32_t argCount, ...);
+
+/*  Internal string data logging function
+ *
+ *  This should be used to log string data.
+ *
+ *  @param format    printf like format
+ *  @param ...       variable arguments related to format
+ */
+#if defined(__GNUC__) || defined(__CC_ARM)
+void log_str_data(const char *format, ...) __attribute__ ((__format__(__printf__, 1, 2)));
+#else
+void log_str_data(const char *format, ...);
+#endif
+
+/*  Internal string data logging function
+ *
+ *  This should be used to log string data like vprintf.
+ *
+ *  @param format    vprintf like format
+ *  @param args      variable arguments
+ */
+void log_str_vdata(const char *format, va_list args);
+
+/*  Logging lock
+ *
+ *  Thread safety is achieved using mutex lock/unlock, lock is applied in each log call.
+ *  If external library needs to synchronize with mbed logging, set the locking functions
+ *  in `mbed_log_set_mutex_wait` and `mbed_log_set_mutex_release`
+ */
+void mbed_log_lock(void);
+
+/*  Logging unlock
+ *
+ *  Thread safety is achieved using mutex lock/unlock, lock is applied in each log call.
+ *  If external library needs to synchronize with mbed logging, set the locking functions
+ *  in `mbed_log_set_mutex_wait` and `mbed_log_set_mutex_release`
+ */
+void mbed_log_unlock(void);
+
+/*  Logging unlock all
+ *
+ *  Few helper functions passed as argument in Logging API's need additional lock
+ *  which should be released only when entire format string is passed. Unlock all
+ *  does unlocking of all locks using a counter.
+ */
+void mbed_log_unlock_all(void);
+
+typedef void (*mbed_log_mutex_fptr)(void);
+/*  Logging Custom lock
+ *
+ *  Function pointer to application lock can be used instead of mbed logging libraries
+ *  default lock
+ *
+ *  @param  mutex_wait_f  Function pointer to mutex wait function
+ */
+void mbed_log_set_mutex_wait(mbed_log_mutex_fptr mutex_wait_f);
```

Comment:
There's a bunch of declaration here that looks like public API's and would benefits to have doxygen comments as well as being into `mbed_logger.h`.

Reply:
None of them is public API for now, they are added to support mbed_trace library and might change in future. However we might make them public in future.